[SanitizerCoverage] Add gated tracing callbacks support to trace-cmp

thetruestblue opened the pull request:
The option -sanitizer-coverage-gated-trace-callbacks gates the invocation of the trace-pc-guard callbacks based on the value of a global variable, which is stored in a specific section.
In this commit, we extend this feature to trace-cmp and gate the cmp callbacks to the same variable used for trace-pc-guard.

Update SanitizerCoverage doc with this flag.

rdar://135404160

Patch by: Andrea Fioraldi

vitalybuka commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -816,12 +821,30 @@ Value *ModuleSanitizerCoverage::CreateFunctionLocalGateCmp(IRBuilder<> &IRB) {
   return Cmp;
 }
 
+Instruction *ModuleSanitizerCoverage::CreateGateBranch(Function &F,
```
It would be nice to have a separate NFC patch for extracting `CreateGateBranch` method.

vitalybuka commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -1013,19 +1051,10 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
-      // Set the branch weights in order to minimize the price paid when the
-      // gate is turned off, allowing the default enablement of this
-      // instrumentation with as little of a performance cost as possible
-      auto Weights = MDBuilder(*C).createBranchWeights(1, 100000);
```
This can be converted to `auto Weights = MDBuilder(*C).createUnlikelyBranchWeights();`
In a separate patch if possible.

wrotki replied:
+1

`MDNode *MDBuilder::createUnlikelyBranchWeights() {
  // Value chosen to match UR_NONTAKEN_WEIGHT, see BranchProbabilityInfo.cpp
  return createBranchWeights(1, (1U << 20) - 1);
}`

thetruestblue committed:
[SanitizerCoverage] Add gated tracing callbacks support to trace-cmp

The option -sanitizer-coverage-gated-trace-callbacks gates the invocation of the
trace-pc-guard callbacks based on the value of a global variable, which is stored
in a specific section.
In this commit, we extend this feature to trace-cmp and gate the cmp callbacks to
the same variable used for trace-pc-guard.

Update SanitizerCoverage doc with this flag.

rdar://135404160

Patch by: Andrea Fioraldi

thetruestblue committed:
Merge branch 'main' into sancovtracecallbacks

thetruestblue committed:
Update SanitizerCoverage.cpp

thetruestblue committed:
Merge branch 'main' into sancovtracecallbacks
